### ISSUE TITLE
Panic "Ran out of registers in GPR" when compiling a function

### ISSUE THREAD

**Issue by bnjbvr** [CraneStation/cranelift#498]
When running the following function with the following settings, I get a register allocation crash:

```
function u0:3203(i32, i32, i32, i32, i32, i32, i64 vmctx) -> i32 uext baldrdash {
    gv0 = vmctx
    gv1 = load.i64 notrap aligned gv0
    heap0 = static gv1, min 0x0300_0000, bound 0x0001_0000_0000, guard 0x8000_0000, index_type i32

                                ebb0(v0: i32, v1: i32, v2: i32, v3: i32, v4: i32, v5: i32, v6: i64):
@0001                               v8 = iconst.i32 0
@0007                               v10 = heap_addr.i64 heap0, v2, 1
@0007                               v11 = load.i32 v10
@000e                               v12 = heap_addr.i64 heap0, v1, 1
@000e                               v13 = load.i32 v12
@0013                               v14 = icmp ult v11, v13
@0013                               v15 = bint.i32 v14
@0020                               v16 = heap_addr.i64 heap0, v0, 1
@0020                               v17 = load.i32 v16
@002d                               v18 = icmp ult v13, v17
@002d                               v19 = bint.i32 v18
@002e                               brz v19, ebb8
@0032                               brz v15, ebb5
@0038                               v20 = heap_addr.i64 heap0, v0, 1
@0038                               store v11, v20
@003f                               v21 = heap_addr.i64 heap0, v2, 1
@003f                               store v17, v21
@0042                               v22 = iconst.i32 1
@0046                               jump ebb3(v17, v17, v22)

                                ebb5:
@004d                               v23 = heap_addr.i64 heap0, v0, 1
@004d                               store.i32 v13, v23
@0054                               v24 = heap_addr.i64 heap0, v1, 1
@0054                               store.i32 v17, v24
@0059                               v25 = heap_addr.i64 heap0, v2, 1
@0059                               v26 = load.i32 v25
@0064                               v27 = icmp ult v26, v17
@0064                               v28 = bint.i32 v27
@0065                               brz v28, ebb7
@006b                               v30 = heap_addr.i64 heap0, v1, 1
@006b                               store v26, v30
@0072                               v31 = heap_addr.i64 heap0, v2, 1
@0072                               store.i32 v17, v31
@0075                               v32 = iconst.i32 2
@0077                               jump ebb6(v32, v17, v17)

                                ebb7:
@0080                               v33 = iconst.i32 1
@0082                               jump ebb6(v33, v26, v26)

                                ebb6(v29: i32, v60: i32, v71: i32):
@0085                               jump ebb4(v60, v71, v29)

                                ebb8:
@0088                               brz.i32 v15, ebb12
@008e                               v35 = heap_addr.i64 heap0, v1, 1
@008e                               store.i32 v11, v35
@0095                               v36 = heap_addr.i64 heap0, v2, 1
@0095                               store.i32 v13, v36
@009a                               v37 = heap_addr.i64 heap0, v1, 1
@009a                               v38 = load.i32 v37
@00a1                               v39 = heap_addr.i64 heap0, v0, 1
@00a1                               v40 = load.i32 v39
@00a6                               v41 = icmp ult v38, v40
@00a6                               v42 = bint.i32 v41
@00a7                               brz v42, ebb11
@00ad                               v44 = heap_addr.i64 heap0, v0, 1
@00ad                               store v38, v44
@00b4                               v45 = heap_addr.i64 heap0, v1, 1
@00b4                               store v40, v45
@00b7                               v46 = iconst.i32 2
@00bd                               v47 = heap_addr.i64 heap0, v2, 1
@00bd                               v48 = load.i32 v47
@00c2                               jump ebb10(v48, v48, v46)

                                ebb11:
@00c3                               v49 = iconst.i32 1
@00cd                               jump ebb10(v13, v13, v49)

                                ebb10(v43: i32, v73: i32, v78: i32):
@00ce                               jump ebb9(v43, v73, v78)

                                ebb12:
@00cf                               v50 = iconst.i32 0
@00d9                               jump ebb9(v11, v11, v50)

                                ebb9(v34: i32, v72: i32, v77: i32):
@00dc                               jump ebb4(v34, v72, v77)

                                ebb4(v59: i32, v70: i32, v76: i32):
@00dd                               jump ebb3(v59, v70, v76)

                                ebb3(v58: i32, v69: i32, v75: i32):
@00e0                               v56 = heap_addr.i64 heap0, v3, 1
@00e0                               v57 = load.i32 v56
@00e7                               v61 = icmp ult v57, v58
@00e7                               v62 = bint.i32 v61
@00e8                               brz v62, ebb16
@00ee                               v68 = heap_addr.i64 heap0, v2, 1
@00ee                               store v57, v68
                                    v51 -> v3
@00f5                               v74 = heap_addr.i64 heap0, v51, 1
@00f5                               store v69, v74
@00fa                               v79 = iconst.i32 1
@00fc                               v80 = iadd v75, v79
@0101                               v81 = heap_addr.i64 heap0, v2, 1
@0101                               v82 = load.i32 v81
@0108                               v88 = heap_addr.i64 heap0, v1, 1
@0108                               v89 = load.i32 v88
@010d                               v90 = icmp ult v82, v89
@010d                               v91 = bint.i32 v90
                                    v51 -> v3
@010e                               brz v91, ebb14(v51, v80)
@0114                               v92 = heap_addr.i64 heap0, v1, 1
@0114                               store v82, v92
@011b                               v93 = heap_addr.i64 heap0, v2, 1
@011b                               store v89, v93
@0120                               v94 = iconst.i32 2
@0122                               v95 = iadd v75, v94
@0127                               v96 = heap_addr.i64 heap0, v1, 1
@0127                               v97 = load.i32 v96
@012e                               v103 = heap_addr.i64 heap0, v0, 1
@012e                               v104 = load.i32 v103
@0133                               v105 = icmp ult v97, v104
@0133                               v106 = bint.i32 v105
@0134                               brz v106, ebb15(v95)
@013a                               v107 = heap_addr.i64 heap0, v0, 1
@013a                               store v97, v107
@0141                               v108 = heap_addr.i64 heap0, v1, 1
@0141                               store v104, v108
@0146                               v109 = iconst.i32 3
@0148                               v110 = iadd v75, v109
@014b                               jump ebb15(v110)

                                ebb15(v130: i32):
                                    v123 -> v3
@014c                               jump ebb14(v123, v130)

                                ebb14(v122: i32, v129: i32):
                                    v112 -> v4
                                    v138 -> v2
                                    v151 -> v1
                                    v164 -> v0
@014d                               jump ebb13(v112, v122, v129, v138, v151, v164)

                                ebb16:
                                    v113 -> v4
                                    v51 -> v3
                                    v63 -> v2
                                    v83 -> v1
                                    v98 -> v0
@0152                               jump ebb13(v113, v51, v75, v63, v83, v98)

                                ebb13(v111: i32, v121: i32, v128: i32, v137: i32, v150: i32, v163: i32):
@0155                               v119 = heap_addr.i64 heap0, v111, 1
@0155                               v120 = load.i32 v119
@015c                               v124 = heap_addr.i64 heap0, v121, 1
@015c                               v125 = load.i32 v124
@0161                               v126 = icmp uge v120, v125
@0161                               v127 = bint.i32 v126
@0162                               brz v127, ebb17
@0166                               jump ebb1(v128)

                                ebb17:
@016c                               v131 = heap_addr.i64 heap0, v121, 1
@016c                               store.i32 v120, v131
@0173                               v132 = heap_addr.i64 heap0, v111, 1
@0173                               store.i32 v125, v132
@0178                               v133 = iconst.i32 1
@017a                               v134 = iadd.i32 v128, v133
@017f                               v135 = heap_addr.i64 heap0, v121, 1
@017f                               v136 = load.i32 v135
@0186                               v140 = heap_addr.i64 heap0, v137, 1
@0186                               v141 = load.i32 v140
@018b                               v142 = icmp uge v136, v141
@018b                               v143 = bint.i32 v142
@018c                               brz v143, ebb18
@0190                               jump ebb1(v134)

                                ebb18:
@0196                               v144 = heap_addr.i64 heap0, v137, 1
@0196                               store.i32 v136, v144
@019d                               v145 = heap_addr.i64 heap0, v121, 1
@019d                               store.i32 v141, v145
@01a2                               v146 = iconst.i32 2
@01a4                               v147 = iadd.i32 v128, v146
@01a9                               v148 = heap_addr.i64 heap0, v137, 1
@01a9                               v149 = load.i32 v148
@01b0                               v153 = heap_addr.i64 heap0, v150, 1
@01b0                               v154 = load.i32 v153
@01b5                               v155 = icmp uge v149, v154
@01b5                               v156 = bint.i32 v155
@01b6                               brz v156, ebb19
@01ba                               jump ebb1(v147)

                                ebb19:
@01c0                               v157 = heap_addr.i64 heap0, v150, 1
@01c0                               store.i32 v149, v157
@01c7                               v158 = heap_addr.i64 heap0, v137, 1
@01c7                               store.i32 v154, v158
@01cc                               v159 = iconst.i32 3
@01ce                               v160 = iadd.i32 v128, v159
@01d3                               v161 = heap_addr.i64 heap0, v150, 1
@01d3                               v162 = load.i32 v161
@01da                               v166 = heap_addr.i64 heap0, v163, 1
@01da                               v167 = load.i32 v166
@01df                               v168 = icmp uge v162, v167
@01df                               v169 = bint.i32 v168
@01e0                               brz v169, ebb20
@01e4                               jump ebb1(v160)

                                ebb20:
@01ea                               v170 = heap_addr.i64 heap0, v163, 1
@01ea                               store.i32 v162, v170
@01f1                               v171 = heap_addr.i64 heap0, v150, 1
@01f1                               store.i32 v167, v171
@01f6                               v172 = iconst.i32 4
@01f8                               v173 = iadd.i32 v128, v172
@01f9                               jump ebb2(v173)

                                ebb2(v9: i32):
@01fa                               jump ebb1(v9)

                                ebb1(v7: i32):
@01fa                               return v7
}
```

Run with:

```
cargo run compile /tmp/file.clif --set return_at_end --set avoid_div_traps --set baldrdash_prologue_words=3 --set allones_funcaddrs --set probestack_enabled=false --set opt_level=best --target x86_64
```

I get the following panic, with a fresh compiled Cranelift which commit hash is 083d1e83f104f12093313582a56538a3506afb1f:

```
thread 'main' panicked at 'Ran out of registers in GPR', lib/codegen/src/regalloc/coloring.rs:848:21
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:391
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:346
   7: cranelift_codegen::regalloc::coloring::Context::iterate_solution
             at lib/codegen/src/regalloc/coloring.rs:848
   8: cranelift_codegen::regalloc::coloring::Context::visit_inst::{{closure}}
             at lib/codegen/src/regalloc/coloring.rs:423
   9: <core::result::Result<T, E>>::unwrap_or_else
             at /checkout/src/libcore/result.rs:748
  10: cranelift_codegen::regalloc::coloring::Context::visit_inst
             at lib/codegen/src/regalloc/coloring.rs:421
  11: cranelift_codegen::regalloc::coloring::Context::visit_ebb
             at lib/codegen/src/regalloc/coloring.rs:170
  12: cranelift_codegen::regalloc::coloring::Context::run
             at lib/codegen/src/regalloc/coloring.rs:153
  13: cranelift_codegen::regalloc::coloring::Coloring::run
             at lib/codegen/src/regalloc/coloring.rs:138
  14: cranelift_codegen::regalloc::context::Context::run
             at lib/codegen/src/regalloc/context.rs:182
  15: cranelift_codegen::context::Context::regalloc
             at lib/codegen/src/context.rs:306
  16: cranelift_codegen::context::Context::compile
             at lib/codegen/src/context.rs:148
  17: cranelift_codegen::context::Context::compile_and_emit
             at lib/codegen/src/context.rs:101
  18: clif_util::compile::handle_module
             at src/compile.rs:103
  19: clif_util::compile::run
             at src/compile.rs:71
  20: clif_util::main
             at src/clif-util.rs:231
  21: std::rt::lang_start::{{closure}}
             at /checkout/src/libstd/rt.rs:74
  22: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  23: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  24: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/rt.rs:58
  25: std::rt::lang_start
             at /checkout/src/libstd/rt.rs:74
  26: main
  27: __libc_start_main
  28: _start
```

Will try to reduce it...

**Comment by sunfishcode** [comment]
This is the same assertion failure as https://github.com/CraneStation/cranelift/issues/335, which has another testcase which may be of interest.

**Comment by sunfishcode** [comment]
This is fixed in #552!
